io-sim/src/Control/Monad/IOSim.hs

```diff
@@ -751,3 +758,73 @@ compareTracesST (Just passing) trace = do
                 Nothing -> error "compareTraceST: invariant violation"
         wakeup _ _ SimTrace {} = error "compareTracesST: invariant violation"
         wakeup _ _ fail = return fail
+
+--
+-- QuickCheck monadic combinators
+--
+
+
+
+-- | Like <monadicST https://hackage.haskell.org/package/QuickCheck-2.14.3/docs/Test-QuickCheck-Monadic.html#v:monadicST>.
+--
+-- Note: it calls `traceResult` in non-strict mode, e.g. leaked threads do not
+-- cause failures.
+--
+-- @since 1.4.1.0
+--
+monadicIOSim_ :: Testable a
+              => (forall s. PropertyM (IOSim s) a)
+              -> Property
+monadicIOSim_ sim =
+    monadicIOSim
+      (\e -> case traceResult False e of
+          Left e  -> counterexample (show e) False
+          Right p -> p)
+      id
+      sim
+
+-- | A more general version of `monadicIOSim_`, which:
+--
+-- * allows to run in monad stacks build on top of `IOSim`;
+-- * gives more control how to attach debugging information to failed
+--   tests.
+--
+-- Note, to use this combinator your monad needs to be defined as:
+--
+-- > newtype M s a = M s { runM :: ReaderT State (IOSim s) a }
+--
+-- It's important that `M s` is a monad.  For such a monad one you'll need provide
+-- a natural transformation:
+-- @
+--   -- the state could also be created as an `IOSim` computation.
+--   nat :: forall s a. State -> M s a -> 'IOSim' s a
+--   nat state m = runStateT (runM m) state
+-- @
+--
+-- @since 1.4.1.0
+--
+monadicIOSim :: (Testable a, forall s. Monad (m s))
+             => (SimTrace Property -> Property)
+             -- ^ Allows to trace `SimTrace` in counterexamples.  The simplest
+             -- use case is to pass which `monadicIOSim` does.
```

Comment:
```suggestion
             -- use case is to pass which `monadicIOSim_` does.
```

Reply:
Any better idea for the two names `monadicIOSim_` and `monadicIOSim`.  I put them this way because `monadicIOSim` might be actually more useful than the specialised one since it's useful when dealing with monad transformer stacks.